frame/support/procedural/src/pallet/expand/storage.rs

```diff
@@ -94,54 +93,45 @@ pub fn expand_storages(def: &mut Def) -> proc_macro2::TokenStream {
 
 			let ty = match &storage.metadata {
 				Metadata::Value { value } => {
-					let value = clean_type_string(&quote::quote!(#value).to_string());
 					quote::quote_spanned!(storage.attr_span =>
-						#frame_support::metadata::StorageEntryType::Plain(
-							#frame_support::metadata::DecodeDifferent::Encode(#value)
+						#frame_support::metadata::v13::StorageEntryType::Plain(
+							#frame_support::scale_info::meta_type::<#value>()
 						)
 					)
 				},
 				Metadata::Map { key, value } => {
-					let value = clean_type_string(&quote::quote!(#value).to_string());
-					let key = clean_type_string(&quote::quote!(#key).to_string());
 					quote::quote_spanned!(storage.attr_span =>
-						#frame_support::metadata::StorageEntryType::Map {
+						#frame_support::metadata::v13::StorageEntryType::Map {
 							hasher: <#full_ident as #metadata_trait>::HASHER,
-							key: #frame_support::metadata::DecodeDifferent::Encode(#key),
-							value: #frame_support::metadata::DecodeDifferent::Encode(#value),
+							key: #frame_support::scale_info::meta_type::<#key>(),
+							value: #frame_support::scale_info::meta_type::<#value>(),
 							unused: false,
 						}
 					)
```

Comment:
I think we can now merge: `StorageValueMetadata`, `StorageMapMetadata`, `StorageDoubleMapMetadata` into:
```rust
trait StorageEntryMetadata {
fn ty() -> StorageEntryType;
fn default() -> Vec<u8>;
const MODIFIER: StorageEntryModifier;
const NAME: &'static str;
}
```

But we can do in a follow up PR

Reply:
I might do this now because I need to remove `DefaultByteGetter` which lives in the legacy `frame-metadata` crate, and I'm about to remove it.

Reply:
Just started this: what about the `HASHER` in `map` and the `HASHER1`, `HASHER2` in `doublemap`?

Reply:
at the end in the metadata we only need `ty`, `default`, `name` and `modifier`.
`HASHER`, `HASHER1` and `HASHER2` were only needed in order to construct `ty`, now `ty` can be entirely constructed without the macro help (the macro was needed before to copy/paste the types), so we can remove those constants.

`StorageEntryMetadata`  is implemented in rust in frame/support/src/storage/types/* for each kinds, and macro use the trait to generate the metadata for `ty`, `default`, `modifier` and `name`.
The macro is only parsing the storage item to get the docs to put in metadata.